discovery-infra/test_infra/helper_classes/kube_helpers.py

```diff
@@ -0,0 +1,456 @@
+need to handle registration and deregistration in assisted-service.
+For example:
+
+with cluster_deployment_context(kube_api_client) as cluster_deployment:
+    print(cluster_deployment.status()) 
+
+When a ClusterDeployment has sufficient data, installation will be started 
+automatically.
+"""
+
+
+# silence kubernetes debug messages.
+logging.getLogger('kubernetes').setLevel(logging.INFO)
+
+logger = logging.getLogger(__name__)
+
+
+def create_api_client(kubeconfig_path: Optional[str] = None) -> ApiClient:
+    logger.info('creating kube client with config file: %s', kubeconfig_path)
+
+    conf = Configuration()
+    load_kube_config(config_file=kubeconfig_path, client_configuration=conf)
+    return ApiClient(configuration=conf)
+
+
+def _does_string_contain_value(s: Union[str, None]) -> bool:
+    return s and s != '""'
+
+
+class ObjectReference(dict):
+    """
+    A dict that contains the information required to to let you locate a
+    referenced Kube API resource.
+    """
+
+    def __init__(self, name: str, namespace: str):
+        dict.__init__(self, name=name, namespace=namespace)
+
+
+class Secret:
+    """
+    A Kube API secret resource that consists of the pull secret data, used
+    by a ClusterDeployment CRD.
+    """
+
+    def __init__(
+            self,
+            api_client: ApiClient,
+            name: str,
+            namespace: str = env_variables['namespace']
+    ):
+        self.v1_api = CoreV1Api(api_client)
+        self.__ref = ObjectReference(name=name, namespace=namespace)
+
+    @property
+    def ref(self):
+        return self.__ref
+
+    def create(self, pull_secret: str = env_variables['pull_secret']) -> None:
+        self.v1_api.create_namespaced_secret(
+            body={
+                'apiVersion': 'v1',
+                'kind': 'Secret',
+                'metadata': self.ref,
+                'stringData': {'pullSecret': pull_secret}
+            },
+            namespace=self.ref['namespace']
+        )
+
+        logger.info('created secret %s', self.ref)
+
+    def delete(self) -> None:
+        self.v1_api.delete_namespaced_secret(
+            name=self.ref['name'],
+            namespace=self.ref['namespace']
+        )
+
+        logger.info('deleted secret %s', self.ref)
+
+
+class _BaseCustomResource(abc.ABC):
+    """
+    Base class for all CRDs, enforces basic methods that every resource must
+    have e.g create, delete, get and status.
+    """
+
+    def __init__(self,  name: str, namespace: str):
+        self.__ref = ObjectReference(name=name, namespace=namespace)
+
+    @property
+    def ref(self):
+        return self.__ref
+
+    @abc.abstractmethod
+    def create(self, **kwargs) -> None:
+        raise NotADirectoryError()
+
+    @abc.abstractmethod
+    def get(self) -> dict:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def delete(self) -> None:
+        raise NotADirectoryError()
```

Comment:
why using ``NotADirectoryError``?
btw I think a simple ``pass`` is enough. The ``abc.abstractmethod`` is enough to make subclasses uninstantable without providing implementation

Reply:
ops, thanks for catching that